### DISCUSSION TITLE
Help to compute MODE statistics with binary of ensemble forecast

### DISCUSSION THREAD

**Discussion by gbodeImoleayo** [dtcenter/METplus#1035] [File I/O] [answered]
Dear All,
I use met tool to compute MODE statistics from raw ensemble mean forecast.

Please I would like to get your assistance on how to compute the MODE statistics with binaries of ensemble forecast instead of using the raw data.

Thank you.

Best regards,
Imoleayo

**Comment by JohnHalleyGotway** [comment]
Imoleayo, I see you have a question about running the MODE tool. It sounds like you've been able to run it on ensemble mean forecast data. And now you have a question about running it with "binaries" of ensemble forecast. However, I don't really understand what you're asking. Can you please clarify?

By "binary", do you mean that each ensemble forecast member's output is stored in a "binary file format"? If so, and that binary format is GRIB1 or GRIB2, then MODE would be able to read it fine. If it's another binary file format that MODE cannot read, then you could write a python embedding script to read that data and serve it up in memory.

But hopefully you can clarify your question some more, and ideally provide sample data to demonstrate.

**Comment by gbodeImoleayo** [comment]
Dear John,Many thanks for your swift response. Sorry, for the unclear explanation. Yes, I have been able to run the MODE Tool on the ensemble mean forecast data.
By Binary, I mean the ensemble mean of 0s and 1s data array after applying the parameter (in this case precipitation) thresholds on the raw individual ensemble member data. For example, let's assume I want to use precipitation threshold of 5mm to identify the object. Firstly, I want to apply this threshold on the raw data of the individual ensemble members to get the 0s and 1s (binaries). Next step is to compute the ensemble mean of the binaries. Lastly, I would like to apply the MODE tool on the binaries to compute the statistics and delineate the object.
Thank you.
Best regards,Imoleayo

**Comment by JohnHalleyGotway** [comment]
Thanks for clarifying. I understand your question much better now. What you are describing, computing the frequency at which ensemble members exceed some threshold, is supported in MET in the ensemble_stat tool.

You can list whichever variables and levels you’d like in the “ens” dictionary of the ensemble_stat config file. And for each, you can specify one or more categorical thresholds (cat_thresh). In the “ensemble_flag” dictionary, be sure that “frequency” is set to true, and then you should see those ensemble relative frequency fields written to the NetCDF output file. Each frequency value is a number between 0 and 1 and gives the rate at which the event (as defined by threshold) is occurring across the members.

Here’s a link to the documentation for that tool:
 https://met.readthedocs.io/en/latest/Users_Guide/ensemble-stat.html
 
Then you can run the NetCDF output from ensmble_stat as input to mode.

**Comment by gbodeImoleayo** [comment]
Dear John,Many thanks for your support and the instructions. I will try this and revert to you.Please I would like to know how to then modify the below section of "MODEConfig_default" file after supplying the binary ensemble mean generated by the ensemble_stat tool.
#####
//

// Forecast and observation fields to be verified

//

fcst = {

   field = {

      name  = "precip_tp";

      level = "(*,*,*)";

   }




   censor_thresh      = [];

   censor_val         = [];

   conv_radius        = 60.0/grid_res; // in grid squares

   conv_thresh        = >=5.0;

   vld_thresh         = 0.5;

   filter_attr_name   = [];

   filter_attr_thresh = [];

   merge_thresh       = >=1.25;

   merge_flag         = THRESH;

}

obs = fcst;
##### 
Thank you.
Best regards,Imoleayo    On Tuesday, July 27, 2021, 08:03:28 PM GMT+2, johnhg ***@***.***> wrote:  
 
 


Thanks for clarifying. I understand your question much better now. What you are describing, computing the frequency at which ensemble members exceed some threshold, is supported in MET in the ensemble_stat tool.

You can list whichever variables and levels you’d like in the “ens” dictionary of the ensemble_stat config file. And for each, you can specify one or more categorical thresholds (cat_thresh). In the “ensemble_flag” dictionary, be sure that “frequency” is set to true, and then you should see those ensemble relative frequency fields written to the NetCDF output file. Each frequency value is a number between 0 and 1 and gives the rate at which the event (as defined by threshold) is occurring across the members.

Here’s a link to the documentation for that tool:
https://met.readthedocs.io/en/latest/Users_Guide/ensemble-stat.html

Then you can run the NetCDF output from ensmble_stat as input to mode.

**Reply by JohnHalleyGotway** [reply]
In MET config files, when reading data from NetCDF files, you set the "name" equal to the NetCDF variable containing the data you want, and you set the "level" string to define how to extract a 2D slice of data from that NetCDF variable.

Here's an example of an ensemble frequency you might find in the ensemble_stat NetCDF output file.
```
	float APCP_24_A24_ENS_FREQ_gt0.0(lat, lon) ;
		APCP_24_A24_ENS_FREQ_gt0.0:name = "APCP_24_ENS_FREQ_gt0.0" ;
		APCP_24_A24_ENS_FREQ_gt0.0:long_name = "APCP_24 at A24 Ensemble Relative Frequency" ;
		APCP_24_A24_ENS_FREQ_gt0.0:level = "A24" ;
		APCP_24_A24_ENS_FREQ_gt0.0:units = "kg/m^2" ;
		APCP_24_A24_ENS_FREQ_gt0.0:_FillValue = -9999.f ;
		APCP_24_A24_ENS_FREQ_gt0.0:init_time = "20091231_120000" ;
		APCP_24_A24_ENS_FREQ_gt0.0:init_time_ut = "1262260800" ;
		APCP_24_A24_ENS_FREQ_gt0.0:valid_time = "20100101_120000" ;
		APCP_24_A24_ENS_FREQ_gt0.0:valid_time_ut = "1262347200" ;
		APCP_24_A24_ENS_FREQ_gt0.0:accum_time = "240000" ;
		APCP_24_A24_ENS_FREQ_gt0.0:accum_time_sec = 86400 ;
```
To read this data into MODE, you'd just use the following in the MODE configuration file:
```
name  = "APCP_24_A24_ENS_FREQ_gt0.0";
level = "(*,*)";
```

**Comment by gbodeImoleayo** [comment]
Ok. Thanks.I will try this and let you know the outcome.
Thank you.
Best regards,Imoleayo

**Comment by gbodeImoleayo** [comment, marked as answer]
Dear John,Again, thanks for your support.I have been able to run the ensemble_stat tool but I couldn't find the binary (0s and 1s) variable in the “ensemble_stat_20170823_030000V_ens.nc“ file (please see attachment). Also, I have attached my ensemble_stat configure file for your review.
Thank you.
Best regards,Imoleayo

**Reply by JohnHalleyGotway** [reply]
Two things here...

1. Please note that while you may have attached something to your email, it did not come through via GitHub discussions. In order to add an attachment please do so directly on the discussions webpage, rather than responding to this email:
   https://github.com/dtcenter/METplus/discussions/1035

2. Look in the Ensemble-Stat NetCDF output file for variables that include "ENS_FREQ" in their name. These variables should contain values in the range [0, 1]... where the value indicates the frequency of the threshold event's occurrence among the input ensemble members at that point. You should see an "ENS_FREQ" variable in the output for combination of "field" and "cat_thresh" specified in the "ens" dictionary of the Ensemble-Stat config file.

For example, take a look at this Ensemble-Stat config file:
https://github.com/dtcenter/MET/blob/833499211e023508829ee2723b900fc20df36236/test/config/EnsembleStatConfig#L41

In the "ens" dictionary, there are 2 thresholds for 2-m Temperature and 1 threshold for 24-hour accumulated precip. So we should see 3 "ENS_FREQ" variables in the output.

**Reply by gbodeImoleayo** [reply]
Dear John,
Thanks for your message. Although a similar variable ("stratiform_rainfall_flux_1_all_all_ENS_FREQ_ge5.0") exist in the output, the values are not 0s and 1s. Here is the attachment contain the “ensemble_stat_20170823_030000V_ens.nc“ and configure files 
Thank you.
[ensemble_stat_files.zip](https://github.com/dtcenter/METplus/files/6951706/ensemble_stat_files.zip)

**Reply by JohnHalleyGotway** [reply]
OK, I see the data you sent and this is exactly the type of output I expect from Ensemble-Stat.

The variable named "stratiform_rainfall_flux_1_all_all_ENS_FREQ_ge5.0" contains a field of values between 0 and 1. For each grid point, the value indicates the frequency of the event among the 18 input ensemble members. So 0.33 means that 6 of 18 members have precip >= 5.0 occurring at that grid point.

Ensemble-Stat does NOT write fields of 0's and 1's separately for each ensemble member. Is that really what you're looking for? Your original question was about running MODE on a binary field derived from an ensemble. So how do you want to define that 0/1 binary field?

If it were me, I'd read that "stratiform_rainfall_flux_1_all_all_ENS_FREQ_ge5.0" variable directly into MODE. And then in the MODE config file, I'd define a convolution radius (i.e. smoothing radius) of 0 since the data is already pretty smooth. And I'd define a convolution threshold as something between 0 and 1. For example:
```
conv_radius = 0;
conv_thresh = >0;
```
This would define an object as being anywhere the event (precip >= 5.0) was occurring in AT LEAST one ensemble member. Or you could increase the conv_thresh to consider only those points where the event occurred more frequently. conv_thresh = >=0.5 would define objects for events occurring in at least 9 of the 18 ensemble members.

Hope that helps clarify.

**Reply by gbodeImoleayo** [reply]
Hi John, thanks for your suggestions. It is very helpful.
Following this suggestion, I also tried to create similar "stratiform_rainfall_flux_1_all_all_ENS_FREQ_ge5.0" variable using ensemble_stat for the observation so that I can use both the ensemble output and observation output as inputs to MODE. But I get the below error message. 

###MODE Command issued
mode ../ensemble_stat/ens_output/ensemble_stat_20170823_000000V_ens.nc ../ensemble_stat/ens_output_obs/ensemble_stat_20170823_000000V_ens.nc MODEConfig_modified -outdir metum_out -v 2
###

Here is the MODE config file.
[MODEConfig_modified.txt](https://github.com/dtcenter/METplus/files/7019245/MODEConfig_modified.txt)


#####Error Message
DEBUG 1: Default Config File: /usr/local/share/met/config/MODEConfig_default
DEBUG 1: Match Config File: MODEConfig_modified
DEBUG 1: Merge Config File: MODEConfig_modified
DEBUG 1: Forecast File: ../ensemble_stat/ens_output/ensemble_stat_20170823_000000V_ens.nc
DEBUG 1: Observation File: ../ensemble_stat/ens_output_obs/ensemble_stat_20170823_000000V_ens.nc
ERROR  :
ERROR  : setup_fcst_obs_data() -> can't get forecast data "stratiform_rainfall_flux_3_all_all_ENS_NMEP_ge5.0_NBRHD25_GAUSSIAN1(*,*)" from file "../ensemble_stat/ens_output/ensemble_stat_20170823_000000V_ens.nc"
ERROR  :
HDF5-DIAG: Error detected in HDF5 (1.8.12) thread 0:
  #000: ../../src/H5T.c line 1763 in H5Tclose(): not a datatype
    major: Invalid arguments to routine
    minor: Inappropriate type
NetCDF: HDF error
file: ncFile.cpp  line:33
HDF5-DIAG: Error detected in HDF5 (1.8.12) thread 0:
  #000: ../../src/H5T.c line 1763 in H5Tclose(): not a datatype
    major: Invalid arguments to routine
    minor: Inappropriate type
NetCDF: HDF error
file: ncFile.cpp  line:33
#####

Is this error related to the input file format?

Thank you.

**Reply by JohnHalleyGotway** [reply]
Please run 'ncdump -h' on that Ensemble-Stat output file to confirm that that variable name actually exists:
```
ncdump -h ../ensemble_stat/ens_output/ensemble_stat_20170823_000000V_ens.nc | grep stratiform_rainfall_flux_3_all_all_ENS_NMEP_ge5.0_NBRHD25_GAUSSIAN1
```
If not, correct the variable name in the MODE config file to match what actually exists in the Ensemble-Stat output file.
If so, please gzip that Ensemble-Stat file and attach it to this issue so that I can look more closely at it:
```
gzip ../ensemble_stat/ens_output/ensemble_stat_20170823_000000V_ens.nc
```

**Reply by gbodeImoleayo** [reply]
Hi John, Thanks for your reply and observations.
Using the "ncdump -h ..." command, I discovered that the variable name does not exist in the file.
 
After checking this issue closely,  I then realised that the number (e.g. 3 in  stratiform_rainfall_flux_3_all_all_ENS_NMEP_ge5.0_NBRHD25_GAUSSIAN1) variable name varies because I generated the ensemble_stat files using a loop from 0 to 24. This means that each of the ensemble_stat file has an index number different from the previous file.  

Sorry, this is the cause of the problem and I have fixed it. Now the MODE tool is working fine.

Thank you.

**Reply by gbodeImoleayo** [reply]
Hi John, here is an output from the mode. 
[mode_000000L_20170825_180000V_NAA_obj.nc.gz](https://github.com/dtcenter/METplus/files/7028234/mode_000000L_20170825_180000V_NAA_obj.nc.gz)

I used the MODE configure file 
[MODEConfig_modified_tmp.txt](https://github.com/dtcenter/METplus/files/7028226/MODEConfig_modified_tmp.txt)

The output is not providing desirable results. Please could you kindly suggest how to set the MODE parameter in the config file to get a reasonable output?

Thank you.

**Reply by JohnHalleyGotway** [reply]
Thanks for sending this MODE output data to illustrate the situation.

Here are plots of the raw forecast and observation fields of data:
```
plot_data_plane mode_000000L_20170825_180000V_NAA_obj.nc fcst.ps 'name="fcst_raw"; level="(*,*)";'
plot_data_plane mode_000000L_20170825_180000V_NAA_obj.nc obs.ps 'name="obs_raw"; level="(*,*)";'
```
<img width="1120" alt="Screen Shot 2021-08-23 at 9 09 54 AM" src="https://user-images.githubusercontent.com/21087144/130471619-11198720-f364-4db7-bd53-5bc0f3894313.png">

Note that the values in both fields range from 0.0 up to about 1.0. And you've configured MODE to define objects by doing no smoothing (i.e. conv_radius = 0;) and then thresholding the result (i.e. conv_thresh = >=0;). But all the values ARE greater-than-or-equal-to 0, which means that every grid point is included in the object. Instead, switch to using greater-than 0, as shown below.
```
   conv_radius        = 0;
   conv_thresh        = >0;
```
And you could consider increasing that threshold (i.e. >=0.25) if you want to require that more ensemble members agree to justify creating an object.

Hope that helps clarify.

**Reply by gbodeImoleayo** [reply]
Hi John,
Thanks for the suggestion, your time and the plots. Yes, It helps a lot.

I have tried the conv_thresh >=0.25.

Please see the output below. As you would see there is a large difference between metum ensemble mean and wrf deterministic forecast. I guess this has to do with the way the frequency variable was created from ensemble_stat for the observation and WRF.
[mode_output.tar.bz2.gz](https://github.com/dtcenter/METplus/files/7035875/mode_output.tar.bz2.gz)

I think it is also important to have the observation, metum and wrf frequency data on the same scale before applying the mode tool. 
Please what is your opinion on this?

Thank you.

**Reply by JohnHalleyGotway** [reply]
Sorry. I lost track of this METplus discussion. In general, I do always recommend that the forecast and observation data be placed on the same "scale" prior to comparing them.

Looking in the MODE NetCDF output files you sent, I see the following:
- The wrfd3 forecast data contains values of 0 or 1.
- The metum forecast data contains values BETWEEN 0 and 1.
- The observation data for wrfd3 and metum contains values of 0 or 1.

MODE defines objects from the input data based on what features you, as the user, would like to study. When the input field is a bitmap of only 0's/1's, you don't really have a choice. You're pretty limited in how you define objects. For the metum data with values between 0 and 1 you can pick how you'd like to define objects. This is entirely up to you and should be based on the scientific questions you're studying. I really can't tell you what to do here.

**Comment by gbodeImoleayo** [comment]
Many thanks for your reply. You are very correct that the choice of threshold is limited for data with 0s or 1s. In such case, one would only get a reasonable result when the threshold for the objects is set to greater than 0, which will logically place data with range 0 to 1 on the same scale as the 0s or 1s.Thank you.

**Reply by JohnHalleyGotway** [reply]
Thanks for confirming. At this point, I'm going to "Lock" this conversation which will prevent future posts on this Discussion topic. Our team has decided to do this once discussions come to a logical conclusion. That helps us keep track of which discussions are still open and require attention. It also helps to keep each discussion as focussed as possible.

But if more issues or questions arise in your use of the METplus software, please feel free to create a new discussion.
